### PR TITLE
improve(insured-bridge-relayer): If deposit cannot be found for relay, conduct another block search using relay.quoteTime

### PR DESCRIPTION
**Pull request by nicholaspai**
**Motivation**

Currently, the disputer will submit disputes on relays where it cannot find a matching deposit for. There is an unlikely case where the deposit event cannot be found in the L2 client's block search because a lot of time has passed between the deposit and relay transactions. In this case, the disputer might not see a valid deposit for a given relay and submit a false dispute.

This PR adds a feature where the bot, upon not finding a deposit using the L2 client's default blocks search, uses the relay's `quoteTime` to conduct a second block search to find the deposit. If it can't find the deposit again, then we can be more confident that the relay has no matching deposit on L2. 

The downside to this is that there is a potential additional call to `findBlockNumberAtTimestamp` which is an expensive function.

**Summary**

Briefly summarize what changes were made to accomplish the motivation above.

**Details**

This may be unnecessary for some PRs. Catch-all for detailed explanations about the implementation decisions and implications of the change.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [X]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

### CODE REVIEW

File: packages/insured-bridge-relayer/src/Relayer.ts [hunk 1]
```diff
@@ -721,4 +725,56 @@ export class Relayer {
...
+  // Return unique deposit event matching relay
+  private async matchRelayWithDeposit(relay: Relay): Promise<Deposit | undefined> {
+    // First try to fetch deposit from the L2 client's default block search config. This should work in most cases.
+    let deposit: Deposit | undefined = this.l2Client.getDepositByHash(relay.depositHash);
+
+    // We could not find a deposit using the L2 client's default block search config. Next, we'll modify the block
+    // searcg config using the relay's quote time. This allows us to capture any deposits that happened outside of
+    // the L2 client's default block search config.
+    if (deposit === undefined) {
+      const l2BlockForDepositQuoteTime = await findBlockNumberAtTimestamp(this.l2Client.l2Web3, relay.quoteTimestamp);
```
[the remark this answers is not in the document]
Comment (chrismaree) [reply packages/insured-bridge-relayer/src/Relayer.ts:738]: for now this is likely fine though as in almost all cases this branch won't get hit

File: packages/insured-bridge-relayer/src/Relayer.ts [hunk 2]
```diff
@@ -721,4 +725,56 @@ export class Relayer {
         break;
     }
   }
+
+  // Return unique deposit event matching relay
+  private async matchRelayWithDeposit(relay: Relay): Promise<Deposit | undefined> {
+    // First try to fetch deposit from the L2 client's default block search config. This should work in most cases.
+    let deposit: Deposit | undefined = this.l2Client.getDepositByHash(relay.depositHash);
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply packages/insured-bridge-relayer/src/Relayer.ts:732]: Yeah really good point and optimization will fix

File: packages/insured-bridge-relayer/src/Relayer.ts [hunk 3]
```diff
@@ -721,4 +725,56 @@ export class Relayer {
...
+
+    // We could not find a deposit using the L2 client's default block search config. Next, we'll modify the block
+    // searcg config using the relay's quote time. This allows us to capture any deposits that happened outside of
+    // the L2 client's default block search config.
+    if (deposit === undefined) {
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply packages/insured-bridge-relayer/src/Relayer.ts:737]: Fixed!

File: packages/insured-bridge-relayer/src/Relayer.ts [hunk 4]
```diff
@@ -181,7 +184,7 @@ export class Relayer {
       const relayExpired = await this.isRelayExpired(relay, relay.l1Token);
       if (relayExpired.isExpired) {
         this.logger.debug({
-          at: "Relayer",
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply packages/insured-bridge-relayer/src/Relayer.ts:184]: Yeah good point let's do that

File: packages/insured-bridge-relayer/src/Relayer.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (nicholaspai) [reply packages/insured-bridge-relayer/src/Relayer.ts:750]: I think its safer to use the contract's deployment time but one problem with this is that the max amount of blocks that we can query might be < `now - contract.deploymentTime` which means we'd need to add additional logic to break up this time period into chunks. Looking up the relay.quoteTime seems like it'd capture the deposit in all cases since the deposit box forces `quoteTime` to be +/- 10 minutes from the current time. So as long as the lookback is > 20 minutes long, it will capture it.

File: packages/insured-bridge-relayer/src/Relayer.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (nicholaspai) [reply packages/insured-bridge-relayer/src/Relayer.ts:184]: Done

File: packages/insured-bridge-relayer/src/Relayer.ts [hunk 5]
```diff
@@ -30,6 +32,8 @@ export enum RelaySubmitType {
 }
 
 export class Relayer {
+  private readonly l2BlockFinder: BlockFinder<BlockTransactionBase>;
```
Comment (chrismaree) [inline packages/insured-bridge-relayer/src/Relayer.ts:35]: why does this have to be defined with this `BlockTransactionBase`? @mrice32 's PR had the same type declaration; why is this required?

File: packages/insured-bridge-relayer/src/Relayer.ts [hunk 6]
```diff
@@ -50,25 +56,31 @@ export class Relayer {
...
   async checkForPendingDepositsAndRelay(): Promise<void> {
-    this.logger.debug({ at: "Relayer", message: "Checking for pending deposits and relaying" });
+    this.logger.debug({ at: "InsuredBridgeRelayer#Relayer", message: "Checking for pending deposits and relaying" });
 
     // Build dictionary of relayable deposits keyed by L1 tokens. We assume that getRelayableDeposits() filters
     // out Finalized relays.
     const relayableDeposits: RelayableDeposits = this.getRelayableDeposits();
     if (Object.keys(relayableDeposits).length == 0) {
-      this.logger.debug({ at: "Relayer", message: "No relayable deposits for any whitelisted tokens" });
+      this.logger.debug({
```
Comment (chrismaree) [inline packages/insured-bridge-relayer/src/Relayer.ts:72]: thanks for updating these!

File: packages/insured-bridge-relayer/src/Relayer.ts [hunk 7]
```diff
@@ -729,4 +751,71 @@ export class Relayer {
...
+  // Return unique deposit event matching relay
+  private async matchRelayWithDeposit(relay: Relay): Promise<Deposit | undefined> {
+    // First try to fetch deposit from the L2 client's default block search config. This should work in most cases.
+    let deposit: Deposit | undefined = this.l2Client.getDepositByHash(relay.depositHash);
+    if (deposit !== undefined) return deposit;
+    // We could not find a deposit using the L2 client's default block search config. Next, we'll modify the block
+    // search config bridge pool's deployment time. This allows us to capture any deposits that happened outside of
+    // the L2 client's default block search config.
+    else {
+      const bridgePoolDeploymentTime = (
```
Comment (chrismaree) [inline packages/insured-bridge-relayer/src/Relayer.ts:764]: this is great!

File: packages/insured-bridge-relayer/test/Relayer.ts [hunk 8]
```diff
@@ -891,6 +894,84 @@ describe("Relayer.ts", function () {
       await relayer.checkForPendingRelaysAndDispute();
       assert.isTrue(lastSpyLogIncludes(spy, "No pending relays"));
     });
+    it("Before disputing relays for deposits it cannot find, first tries to find deposit in new blocksearch", async function () {
```
Comment (chrismaree) [inline packages/insured-bridge-relayer/test/Relayer.ts:897]: +10

**Review by chrismaree — APPROVED** [review]
This is awesome! we should never need to use this code but in the event, the bots are off for a while and booted up we might need this! good work on adding this. Excellent tests as well.
